### PR TITLE
Implements setReadOnly and clearReadOnly 

### PR DESCRIPTION
**Pull request by mssfang**
added setReadOnly and ClearReadOnly sync and async APIs.
added samples and tests to all APIs.
suppress checkstyle rule for java method naming guideline.

fixes: https://github.com/Azure/azure-sdk-for-java/issues/5356

### CODE REVIEW

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 1]
```diff
@@ -525,13 +525,120 @@
         // Validate that setting and key is not null. The key is used in the service URL so it cannot be null.
         validateSetting(setting);
 
-        return service.delete(serviceEndpoint, setting.getKey(), setting.getLabel(), getETagValue(setting.getETag()),
+        return service.delete(serviceEndpoint, setting.getKey(), setting.getLabel(), null,
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Why did you remove the etag matching here?

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 2]
```diff
@@ -525,13 +525,120 @@
...
+     * value (which will also throw HttpResponseException described below).
+     * @throws IllegalArgumentException If {@code key} is {@code null}.
+     * @throws HttpResponseException If {@code key} is an empty string.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<ConfigurationSetting>> clearReadOnlyWithResponse(ConfigurationSetting setting) {
+        return withContext(context -> clearReadOnly(setting, context));
+    }
+
+    Mono<Response<ConfigurationSetting>> clearReadOnly(ConfigurationSetting setting, Context context) {
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Should this be private

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 3]
```diff
@@ -525,13 +525,120 @@
...
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>Lock the setting with the key-label "prodDBConnection"-"westUS".</p>
+     *
+     * {@codesnippet com.azure.data.appconfiguration.configurationasyncclient.setReadOnly#string-string}
+     *
+     * @param key The key of the configuration setting to add.
+     * @param label The label of the configuration setting to add.
+     * @return The {@link ConfigurationSetting} that was created, if a key collision occurs or the key is an invalid
+     * value (which will also throw HttpResponseException described below).
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Indent continuations

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 4]
```diff
@@ -525,13 +525,120 @@
...
+    /**
+     * Lock the {@link ConfigurationSetting} with a matching key, along with the given label.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>Lock the setting with the key-label "prodDBConnection"-"westUS".</p>
+     *
+     * {@codesnippet com.azure.data.appconfiguration.configurationasyncclient.setReadOnly#string-string}
+     *
+     * @param key The key of the configuration setting to add.
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: The documentation is wrong

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 5]
```diff
@@ -525,13 +525,120 @@
...
+    /**
+     * Unlock the {@link ConfigurationSetting} with a matching key, along with the given label.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>unlock the setting with the key-label "prodDBConnection"-"westUS".</p>
+     *
+     * {@codesnippet com.azure.data.appconfiguration.configurationasyncclient.setReadOnlyWithResponse#ConfigurationSetting}
+     *
+     * @param setting The ConfigurationSetting to unlock.
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: The documentation is incorrect

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 6]
```diff
@@ -525,13 +525,120 @@
...
+     * value (which will also throw HttpResponseException described below).
+     * @throws IllegalArgumentException If {@code key} is {@code null}.
+     * @throws HttpResponseException If {@code key} is an empty string.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<ConfigurationSetting>> setReadOnlyWithResponse(ConfigurationSetting setting) {
+        return withContext(context -> setReadOnly(setting, context));
+    }
+
+    Mono<Response<ConfigurationSetting>> setReadOnly(ConfigurationSetting setting, Context context) {
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Why package private over just private

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 7]
```diff
@@ -525,13 +525,120 @@
...
             null, context)
             .doOnRequest(ignoredValue -> logger.info("Deleting ConfigurationSetting - {}", setting))
             .doOnSuccess(response -> logger.info("Deleted ConfigurationSetting - {}", response.getValue()))
             .doOnError(error -> logger.warning("Failed to delete ConfigurationSetting - {}", setting, error));
     }
 
+    /**
+     * Lock the {@link ConfigurationSetting} with a matching key, along with the given label.
+     *
+     * <p><strong>Code Samples</strong></p>
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: I haven't seen us explicitly declare a section in our java docs as "Code Samples"

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 8]
```diff
@@ -525,13 +525,120 @@
...
+     * @return The {@link ConfigurationSetting} that was created, if a key collision occurs or the key is an invalid
+     * value (which will also throw HttpResponseException described below).
+     * @throws IllegalArgumentException If {@code key} is {@code null}.
+     * @throws HttpResponseException If {@code key} is an empty string.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<ConfigurationSetting> setReadOnly(String key, String label) {
+        return withContext(context -> setReadOnly(
+            new ConfigurationSetting().setKey(key).setLabel(label), context))
+            .flatMap(response -> Mono.justOrEmpty(response.getValue()));
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Why a flatmap here? Can't you just map it?

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 9]
```diff
@@ -525,13 +525,120 @@
...
+
+    Mono<Response<ConfigurationSetting>> setReadOnly(ConfigurationSetting setting, Context context) {
+        // Validate that setting and key is not null. The key is used in the service URL so it cannot be null.
+        validateSetting(setting);
+
+        return service.lockKeyValue(serviceEndpoint, setting.getKey(), setting.getLabel(), null,
+            null, context)
+            .doOnRequest(ignoredValue -> logger.info("Setting read only ConfigurationSetting - {}", setting))
+            .doOnSuccess(response -> logger.info("Set read only ConfigurationSetting - {}", response.getValue()))
+            .doOnError(error -> logger.info("Failed to set read only ConfigurationSetting - {}", setting, error));
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: I dont' think it should be an info level log.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 10]
```diff
@@ -525,13 +525,120 @@
...
+        return withContext(context -> setReadOnly(setting, context));
+    }
+
+    Mono<Response<ConfigurationSetting>> setReadOnly(ConfigurationSetting setting, Context context) {
+        // Validate that setting and key is not null. The key is used in the service URL so it cannot be null.
+        validateSetting(setting);
+
+        return service.lockKeyValue(serviceEndpoint, setting.getKey(), setting.getLabel(), null,
+            null, context)
+            .doOnRequest(ignoredValue -> logger.info("Setting read only ConfigurationSetting - {}", setting))
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Verbose?

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 11]
```diff
@@ -525,13 +525,120 @@
...
+
+    Mono<Response<ConfigurationSetting>> clearReadOnly(ConfigurationSetting setting, Context context) {
+        // Validate that setting and key is not null. The key is used in the service URL so it cannot be null.
+        validateSetting(setting);
+
+        return service.unlockKeyValue(serviceEndpoint, setting.getKey(), setting.getLabel(),
+            null, null, context)
+            .doOnRequest(ignoredValue -> logger.info("Clearing read only ConfigurationSetting - {}", setting))
+            .doOnSuccess(response -> logger.info("Cleared read only ConfigurationSetting - {}", response.getValue()))
+            .doOnError(error -> logger.info("Failed to clear read only ConfigurationSetting - {}", setting, error));
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: I think these log levels should be different for error and on request.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 12]
```diff
@@ -525,13 +525,120 @@
...
+        return withContext(context -> clearReadOnly(setting, context));
+    }
+
+    Mono<Response<ConfigurationSetting>> clearReadOnly(ConfigurationSetting setting, Context context) {
+        // Validate that setting and key is not null. The key is used in the service URL so it cannot be null.
+        validateSetting(setting);
+
+        return service.unlockKeyValue(serviceEndpoint, setting.getKey(), setting.getLabel(),
+            null, null, context)
+            .doOnRequest(ignoredValue -> logger.info("Clearing read only ConfigurationSetting - {}", setting))
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Why doOnRequest over doOnSubscribe?

File: sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationClientJavaDocCodeSnippets.java [hunk 13]
```diff
@@ -278,6 +278,68 @@ public void deleteSetting() {
...
+
+        /**
+         * Generates code sample for using {@link ConfigurationClient#setSettingWithResponse(ConfigurationSetting, Context)}
+         */
+        // BEGIN: com.azure.data.applicationconfig.configurationclient.setReadOnlyWithResponse#ConfigurationSetting-Context
+        Response<ConfigurationSetting> responseSetting = configurationClient
+            .setSettingWithResponse(new ConfigurationSetting().setKey("prodDBConnection").setLabel("westUS"),
+                new Context(key2, value2));
+        System.out
+            .printf("Key: %s, Value: %s", responseSetting.getValue().getKey(), responseSetting.getValue().getValue());
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationClientJavaDocCodeSnippets.java]: This line break is weird. I'd break it after the format string

File: sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationClientJavaDocCodeSnippets.java [hunk 14]
```diff
@@ -278,6 +278,68 @@ public void deleteSetting() {
...
+
+        /**
+         * Generates code sample for using {@link ConfigurationClient#setSettingWithResponse(ConfigurationSetting, Context)}
+         */
+        // BEGIN: com.azure.data.applicationconfig.configurationclient.clearReadOnlyWithResponse#ConfigurationSetting-Context
+        Response<ConfigurationSetting> responseSetting = configurationClient
+            .setSettingWithResponse(new ConfigurationSetting().setKey("prodDBConnection").setLabel("westUS"),
+                new Context(key2, value2));
+        System.out
+            .printf("Key: %s, Value: %s", responseSetting.getValue().getKey(), responseSetting.getValue().getValue());
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationClientJavaDocCodeSnippets.java]: Same. This line break is odd.

File: sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationAsyncClientTest.java [hunk 15]
```diff
@@ -57,8 +57,14 @@ protected void afterTest() {
         logger.info("Cleaning up created key values.");
         client.listSettings(new SettingSelector().setKeys(keyPrefix + "*"))
                 .flatMap(configurationSetting -> {
+                    Mono<Response<ConfigurationSetting>> unlock;
+                    if (configurationSetting.isLocked()) {
+                        unlock = client.clearReadOnlyWithResponse(configurationSetting);
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationAsyncClientTest.java]: You can use a ternary operator.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java (hunk 2 above)
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Sync API will use this clearReadOnly API.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java (hunk 7 above)
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Do you mean it need to change to "Code sample"  as others did 
```
     * <p><strong>Code sample</strong></p>
     * {@codesnippet com.azure.core.http.rest.pagedfluxbase.instantiation}
```

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java (hunk 3 above)
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: value(..)?

File: sdk/appconfiguration/azure-data-appconfiguration/pom.xml [hunk 16]
```diff
@@ -57,7 +57,6 @@
       <groupId>com.azure</groupId>
       <artifactId>azure-core-test</artifactId>
       <version>1.0.0-preview.5</version>
-      <scope>test</scope>
```
Comment (alzimmermsft) [inline sdk/appconfiguration/azure-data-appconfiguration/pom.xml]: This should remain a test dependency

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 17]
```diff
@@ -264,8 +281,7 @@
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<ConfigurationSetting> getSetting(String key, String label, OffsetDateTime asOfDateTime) {
         return withContext(context -> getSetting(new ConfigurationSetting().setKey(key).setLabel(label), asOfDateTime,
-                false, context))
-            .flatMap(response -> Mono.justOrEmpty(response.getValue()));
+                false, context)).map(response -> response.getValue());
```
Comment (alzimmermsft) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: There is a FluxUtil method to handle pulling a response value out, I believe it is called toMono. Should use that in all the places update.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 18]
```diff
@@ -333,7 +349,7 @@
     public Mono<ConfigurationSetting> deleteSetting(String key, String label) {
         return withContext(
             context -> deleteSetting(new ConfigurationSetting().setKey(key).setLabel(label), false, context))
-            .flatMap(response -> Mono.justOrEmpty(response.getValue()));
+            .map(response -> response.getValue());
     }
```
Comment (alzimmermsft) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: This needs to use justOrEmpty as the response value can be null which is an illegal value in a Reactor stream.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 19]
```diff
@@ -82,7 +82,7 @@
      * label is desired.
      * @param value The value associated with this configuration setting key.
      * @return The {@link ConfigurationSetting} that was created, if a key collision occurs or the key is an invalid
-     * value (which will also throw HttpResponseException described below).
+     * value(which will also throw HttpResponseException described below).
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: undo this. the space is necessary.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 20]
```diff
@@ -106,7 +106,7 @@
      *
      * @param setting The setting to add to the configuration service.
      * @return The {@link ConfigurationSetting} that was created, if a key collision occurs or the key is an invalid
-     * value (which will also throw HttpResponseException described below).
+     * value(which will also throw HttpResponseException described below).
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: It feels like you did a Ctrl+R to replace all these instances. Before a parenthesis should be a space in these sentences.

File: sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationAsyncClientTest.java [hunk 21]
```diff
@@ -291,6 +292,108 @@ public void deleteSettingNullKey() {
         assertRunnableThrowsException(() -> client.deleteSettingWithResponse(null, false).block(), NullPointerException.class);
     }
 
+    /**
+     * Tests assert that the setting can not be deleted after lock the setting.
+     */
+    public void setReadOnly() {
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationAsyncClientTest.java]: Are you missing the `@Test` annotation?

Comment (conniey) [reply sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationAsyncClientTest.java]: Same with the ones below.

Comment (conniey) [reply sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationAsyncClientTest.java]: Never mind. There is this weird inheritance thing going on.
